cellars: write tag specific cellars

iMichka opened the pull request:
With #10186 now merged, the tag specific cellar information is being read by brew.
This PR (once merged) will start adding the cellar information for each tag instead
of having a single cellar line on the top of the bottle block.

Each new CI build in homebrew-core will slowly start migrating the cellar lines to
the right place. If keep-old is used, the old "all tag" cellar line is removed and
added to each tag.

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

dawidd6 commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -15,26 +15,12 @@
-    <% end %>
+    <% checksums.each do |checksum| %>
+    <% cellar = checksum["cellar"] %>
+    <%= "sha256" %> "<%= checksum["digest"] %>" => :<%= checksum["tag"] %><% if cellar.is_a? Symbol %>, :cellar => :<%= cellar %><% elsif ![Homebrew::DEFAULT_MACOS_CELLAR, Homebrew::DEFAULT_MACOS_ARM_CELLAR, Homebrew::DEFAULT_LINUX_CELLAR].include?(cellar) && !cellar.blank?%>, :cellar => "<%= cellar %>"<% end %>
```
That's a very long line. Any way we can shrink it a bit?

iMichka replied:
Done, I added the generate_sha256_line method to do just that.

dawidd6 commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -15,26 +15,12 @@
-    <% if ![HOMEBREW_DEFAULT_PREFIX,
-            HOMEBREW_MACOS_ARM_DEFAULT_PREFIX,
-            HOMEBREW_LINUX_DEFAULT_PREFIX].include?(prefix) %>
-    prefix "<%= prefix %>"
```
Are we dropping `prefix` method in bottle blocks?

jonchang replied:
I think `prefix` is gone in favor of `cellar`.

iMichka replied:
Woups I removed the prefix by mistake. The goal this PR is not to touch the prefix part, that can be done in a separate PR. I'll add it back.

MikeMcQuaid reviewed (commented):
Nice work so far @iMichka!

iMichka commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -452,39 +438,44 @@ def parse_json_files(filenames)
-      cellar = bottle_hash["bottle"]["cellar"]
-      cellar = cellar.to_sym if any_cellars.include?(cellar)
-      bottle.cellar cellar
       bottle.prefix bottle_hash["bottle"]["prefix"]
```
As said above I will not modify the prefix part in this PR so leaving this as-is.

iMichka commented on Library/Homebrew/test/dev-cmd/bottle_spec.rb:
```diff
@@ -229,13 +221,18 @@ def stub_hash(parameters)
 end
 
-describe "brew bottle --merge", :integration_test, :needs_linux do
+describe "brew bottle --merge", :integration_test do
```
My bad, I removed it while testing this on Mac. Will add it back.

iMichka commented on Library/Homebrew/software_spec.rb:
```diff
@@ -439,13 +439,13 @@ def checksums
+        "cellar" => collector[tag][:cellar],
       }
     end
-    { sha256: sha256s }
```
This one is tricky.

See https://github.com/Homebrew/brew/blob/f6bc3efac6a5a7f18f241a96572af4cf8eeeb820/Library/Homebrew/dev-cmd/bottle.rb#L207-L210. The bottle block is evaluated there, and checksum is then called in the ERB block: https://github.com/Homebrew/brew/blob/f6bc3efac6a5a7f18f241a96572af4cf8eeeb820/Library/Homebrew/dev-cmd/bottle.rb#L33-L36

I reorganised the output to a structure that I think is clearer, especially now that I added `generate_sha256_line` method.

iMichka replied:
That's basically what's included in this PR. It simplifies the code in the ERB part. We can either open a separate PR for this or include it in this PR.

MikeMcQuaid replied:
> Here's a commit that possibly could be used for a PR: [SeekingMeaning@8693ca4](https://github.com/SeekingMeaning/brew/commit/8693ca4e4)

@SeekingMeaning looks good, opened a PR in https://github.com/Homebrew/brew/pull/10395.

iMichka commented:
Note: this can only be merged once a new release of brew has been made; because the reading part of this needs to be shipped first.

SeekingMeaning commented on Library/Homebrew/software_spec.rb:
```diff
@@ -439,13 +439,30 @@ def checksums
+      if cellar.is_a? Symbol
+        sha256_line += ", :cellar => :#{cellar}"
+      elsif default_cellars.exclude?(cellar)
+        sha256_line += ", :cellar => #{cellar}"
```
```suggestion
        sha256_line += ", :cellar => \"#{cellar}\""
```

iMichka replied:
LGTM, forgot to add those. Thanks.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -478,13 +469,13 @@ def merge(args:)
+        cellar = tag_hash["cellar"]
+        cellar = cellar.to_sym if any_cellars.include?(cellar)
+        sha256_hash = { tag_hash["sha256"] => tag.to_sym, :cellar => cellar }
+        bottle.sha256 sha256_hash
```
Would be great, if possible, to have some tests that touch these lines.

iMichka replied:
These lines are already test by the integration tests we have in the bottle_spec.rb file, under the `describe "brew bottle --merge"`. The `...bottle.json` files contain cellars, and the merge_json_files above makes sure the cellar is not at the right level in the `bottle_hash` hash.

If we want to test this more we need to extract this to a separate function; but I think the current tests are enough here.

MikeMcQuaid replied:
> These lines are already test by the integration tests we have in the bottle_spec.rb file, under the `describe "brew bottle --merge"`.

Ah, it's because the integration test is Linux only. I think it's maybe worth making it on macOS too give this (I appreciate contradicting past me).

> If we want to test this more we need to extract this to a separate function; but I think the current tests are enough here.

Extraction would be fine with me 👍🏻

MikeMcQuaid commented on Library/Homebrew/software_spec.rb:
```diff
@@ -439,13 +439,30 @@ def checksums
+      if cellar.is_a? Symbol
+        sha256_line += ", :cellar => :#{cellar}"
+      elsif default_cellars.exclude?(cellar)
+        sha256_line += ", :cellar => \"#{cellar}\""
```
Would be great, if possible, to have some tests that touch this line.

iMichka replied:
Done

MikeMcQuaid reviewed (approved):
Looks good to me now! More tests would be ideal but aren't essential/blocking. Let's land https://github.com/Homebrew/brew/pull/10395 first.

MikeMcQuaid commented on Library/Homebrew/software_spec.rb:
```diff
@@ -439,13 +439,30 @@ def checksums
+  end
+
+  def generate_sha256_line(tag, digest, cellar)
+    sha256_line = "sha256 \"#{digest}\" => :#{tag}"
+    if cellar.present?
```
```suggestion
    sha256_line = "sha256 \"#{digest}\" => :#{tag}"
    return sha256_line if cellar.blank?

```

Could let you reduce the intention here.

SeekingMeaning commented on Library/Homebrew/software_spec.rb:
```diff
@@ -441,11 +441,29 @@ def checksums
     end
   end
+
+  def generate_sha256_line(tag, digest, cellar)
+    sha256_line = "sha256 \"#{digest}\" => :#{tag}"
+    return sha256_line if cellar.blank?
+
+    default_cellars = [
+      Homebrew::DEFAULT_MACOS_CELLAR,
+      Homebrew::DEFAULT_MACOS_ARM_CELLAR,
+      Homebrew::DEFAULT_LINUX_CELLAR,
+    ]
+    if cellar.is_a? Symbol
+      sha256_line += ", :cellar => :#{cellar}"
+    elsif default_cellars.exclude?(cellar)
+      sha256_line += ", :cellar => \"#{cellar}\""
+    end
+    sha256_line
+  end
```
Here is my proposal for generating `sha256` lines in `dev-cmd/bottle.rb` instead of `software_spec.rb`: https://github.com/SeekingMeaning/brew/commit/50efc59f4

iMichka replied:
Cool, I though about it before, and I went for the easy solution, because I was unsure how to modify `bottle_output` to make that work. Thanks for the help, I made that change too.

SeekingMeaning commented on Library/Homebrew/test/dev-cmd/bottle_spec.rb:
```diff
@@ -229,10 +221,41 @@ def stub_hash(parameters)
+    end
+
+    it "generates a string with non-default cellar path" do
+      expect(homebrew.generate_sha256_line(:catalina, "deadbeef", "/home/test")).to eq(
+        "sha256 \"deadbeef\" => :catalina, :cellar => \"/home/test\"",
+      )
```
```suggestion
      expect(homebrew.generate_sha256_line(:catalina, "deadbeef", "/home/test")).to eq <<~EOS.chomp
        sha256 "deadbeef" => :catalina, :cellar => "/home/test"
      EOS
```

dawidd6 replied:
![3f3w4m](https://user-images.githubusercontent.com/9713907/105617684-7ad88600-5de0-11eb-9362-66495a48d5e2.png)

iMichka replied:
Done

dawidd6 reviewed (approved):
Just tested it locally and it worked just fine. Nice work!

iMichka commented:
Cool. If everyone is fine with the current implementation, I'll merge this just after the next homebrew release, when I have time to monitor the next merges and react to any issues, but I think with all the tests we have in place now this should work out of the box :)

SeekingMeaning commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -542,7 +554,8 @@ def merge_bottle_spec(old_keys, old_bottle_spec, new_bottle_hash)
       new_value = new_values[key].to_s
-      next if key == :cellar && old_value == "any" && new_value == "any_skip_relocation"
+
+      next if key == :cellar
```
```suggestion
```

```patch
- next if key == :sha256
+ next if key == :sha256 || key == :cellar
```

SeekingMeaning commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -450,21 +462,22 @@ def parse_json_files(filenames)
-        end
+    bottles_hashes = {}
+    json_files.each do |json_content|
+      formula_name = json_content.keys.first
+
+      json_hash = json_content[formula_name]
```
```suggestion
      json_content.each_pair do |formula_name, json_hash|
        json_hash = json_hash.deep_dup
```

SeekingMeaning replied:
Actually, we could replace this entire method with this:

```rb
def merge_json_files(json_files)
  json_files.reduce({}) do |hash, json_file|
    json_file.each_value do |json_hash|
      json_bottle = json_hash["bottle"]
      cellar = json_bottle.delete("cellar")
      json_bottle["tags"].each_value do |json_platform|
        json_platform["cellar"] ||= cellar
      end
    end
    hash.deep_merge(json_file)
  end
end
```

SeekingMeaning commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -450,21 +462,22 @@ def parse_json_files(filenames)
+      formula_name = json_content.keys.first
+
+      json_hash = json_content[formula_name]
+      platform = json_hash["bottle"]["tags"].keys.first
+      json_hash["bottle"]["tags"][platform]["cellar"] = json_hash["bottle"]["cellar"]
```
```suggestion
      json_bottle_hash = json_hash["bottle"]
      json_bottle_hash["tags"].each_pair do |platform, json_platform_hash|
        json_platform_hash["cellar"] = json_bottle_hash["cellar"]
      end
```

SeekingMeaning commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -450,21 +462,22 @@ def parse_json_files(filenames)
 
-        second
+      if bottles_hashes.key?(formula_name)
+        bottles_hashes[formula_name]["bottle"]["tags"].merge!(json_hash["bottle"]["tags"])
```
```suggestion
        bottles_hashes[formula_name]["bottle"]["tags"].merge!(json_bottle_hash["tags"])
```

SeekingMeaning commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -450,21 +462,22 @@ def parse_json_files(filenames)
+      if bottles_hashes.key?(formula_name)
+        bottles_hashes[formula_name]["bottle"]["tags"].merge!(json_hash["bottle"]["tags"])
+      else
+        json_hash["bottle"].delete("cellar")
+        bottles_hashes[formula_name] = json_hash
```
```suggestion
        json_bottle_hash.delete("cellar")
        bottles_hashes[formula_name] = json_hash
```

SeekingMeaning commented:
My suggestions in commit form: https://github.com/SeekingMeaning/brew/commit/28bb6cd8d

MikeMcQuaid commented on Library/Homebrew/test/dev-cmd/bottle_spec.rb:
```diff
@@ -227,12 +219,48 @@ def stub_hash(parameters)
+  end
 end
 
 describe "brew bottle --merge", :integration_test, :needs_linux do
```
```suggestion
describe "brew bottle --merge", :integration_test do
```

MikeMcQuaid commented:
💚! @iMichka feel free to merge when you can be around to babysit this 😁

iMichka commented:
Cool, thanks. I'll do that once we have a new release with the reading part, else we will break everything for our users.

SeekingMeaning commented on Library/Homebrew/test/dev-cmd/bottle_spec.rb:
```diff
@@ -227,15 +219,51 @@ def stub_hash(parameters)
     end
   end
+
+  describe "::generate_sha256_line", :needs_linux do
```
Why `needs_linux`?

iMichka replied:
To speed up the Mac tests. Maybe we could introduce: "run_on_linux_to_not_clutter_the_mac_tests" but I think this is fine.

SeekingMeaning replied:
Perhaps I'm wrong but I feel like `generate_sha256_line` shouldn't be that slow; also, not having `needs_linux` makes it easier to test when developing on Mac

iMichka replied:
It's fast. It just the sum of all tests that make things slow. Last time I added tests @MikeMcQuaid asked me to make them linux-only to speed things up. I thought that was a common practice here, but I might be wrong.

SeekingMeaning replied:
Oh, okay. I think he was referring to integration tests being slow, but we can wait for him to clarify

MikeMcQuaid replied:
All integration tests are slow. All non-integration tests are (generally) fast. I think it's fine to remove `:needs_linux` here. Sorry for the confusion!

iMichka replied:
It's Ok, I'll change it. Makes sense.

iMichka commented:
I added one more commit (40ad563279c109753f1e123cfc8eba7e5a9da1be) to skip the "line too long" message for the sha256 lines in bottle blocks and rebased.

Now that brew 2.7.8 is out, I'll merge this one later today and monitor how things go in homebrew-core.

MikeMcQuaid commented:
> I added one more commit ([40ad563](https://github.com/Homebrew/brew/commit/40ad563279c109753f1e123cfc8eba7e5a9da1be)) to skip the "line too long" message for the sha256 lines in bottle blocks and rebased.

Can we wrap this instead e.g. from:

```ruby
sha256 "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur, :cellar => :any_skip_relocation
```

to:

```
sha256 "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur, 
       cellar: :any_skip_relocation
```

MikeMcQuaid commented:
> Can we wrap this instead e.g. from:

Or, to keep things more greppable keep the exception but output:

```ruby
sha256 cellar: :any_skip_relocation, "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur
```

or, if it's not too late to mess around with the format a bit further (as once we start writing this we can't go back really):

```ruby
sha256 tag: :big_sur, cellar: :any_skip_relocation, value: "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f"
```

or something, I dunno.

iMichka commented:
I think being able to grep it is nice, so I would rather keep it on one line. I also like it on one line, it looks nicer to my eye.

Your proposals:
```
sha256 cellar: :any_skip_relocation, "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur
```
This one has the drawback that it might look weird with multiple tags:
```
sha256 cellar: :any_skip_relocation, "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :catalina
sha256 "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur
sha256 cellar: :any, "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :linux
```

```sha256 tag: :big_sur, cellar: :any_skip_relocation, value: "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f"```

I think this might have simplified some code, but only by a few lines here and there. It also means that I need to do yet another PR to read this new format, and let that be pushed to our users with release 3.0.0 first. It's also slightly longer.

From a "legacy syntax" point of view I find that the current implementation that was chosen is the best, as it does not change the DSL that much, it just moves the cellar information to the sha256 line, so it will probably not confuse the maintainer eye that much.

MikeMcQuaid commented:
> This one has the drawback that it might look weird with multiple tags:

True.



> From a "legacy syntax" point of view I find that the current implementation that was chosen is the best, as it does not change the DSL that much, it just moves the cellar information to the sha256 line, so it will probably not confuse the maintainer eye that much.

Yeh, my main issue (that I apologise for not considering before now) is that unless your editor has line wrapping enabled (notably: the GitHub diff UI does not): the `cellar` information no longer visible without scrolling. I don't think this is the end of the world but it is a cause for concern.



> I think this might have simplified some code, but only by a few lines here and there. It also means that I need to do yet another PR to read this new format, and let that be pushed to our users with release 3.0.0 first. It's also slightly longer.

I'd rather hold off on 3.0.0 and get this syntax to be one that we're 100% happy with rather than rush and be stuck with something non-ideal either forever or have to make a similar disruptive change in the medium term. @Homebrew/maintainers what do you think about the above? What's your preferred syntax (of anything suggested or come up with your own version).

dawidd6 commented:
Maybe something simple like:

```ruby
sha256 "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f", :big_sur, :any_skip_relocation
```

I think it's descriptive enough.

So the last argument is always cellar and it's default value could be nil or HOMEBREW_CELLAR or something else.

iMichka commented:
```
sha256 "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f", :big_sur, :any_skip_relocation
```

Is this valid ruby syntax? We must not forget that the part after the sha256 needs to be a hash with key/value pairs.

MikeMcQuaid commented:
> Is this valid ruby syntax?

Yes.



> We must not forget that the part after the sha256 needs to be a hash with key/value pairs.

With the current implementation yes but technically: shouldn't need to be. We could potentially even do something weird like:

```ruby
big_sur cellar: :any_skip_relocation, sha256: "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f"
```

carlocab commented:
I'm partial to anything that puts the actual `sha256` value to the right since it's not really meant for human consumption. I think the keys and values for the `sha256` hash in the bottle block are the wrong way around...

That said, I've also grown accustomed to looking at the right in the bottle block, and that might be a hard habit to break.

iMichka commented:
If we say that we keep the sha256 on the left, I think this excludes the following one:

```
sha256 cellar: :any_skip_relocation, "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur
```

because from a reader's point of view, it feels weird that it starts with sha256, but the sha256 hexdigest is way further to the right. Also, the drawback of this syntax is that for multiple tags, it might look weird:

```
sha256 cellar: :any_skip_relocation, "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :catalina
sha256 "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur
sha256 cellar: :any, "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :linux
```

This leads me to think that the remaining options are (if we do not want to change too many things):

1
```
sha256 "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur, cellar: :any_skip_relocation
```

2
```
sha256 :big_sur => "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f", cellar: :any_skip_relocation
```

Option 1 is the one I propose here and what has been implemented. Option 2 is doable but requires some more changes.

Option 3 is to go more exotic and make more changes, like introducing new "tag" and/or "hexdigest" keys.

carlocab commented:
I actually like Mike's suggestion above:
```ruby
big_sur cellar: :any_skip_relocation, sha256: "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f"
```
It's nice, readable, and puts the most valuable information first.

Of course, this probably requires a whole lot more work...

iMichka commented:
> this probably requires a whole lot more work.

Yes, it does. I know that part of the code now, so I know what needs to be done for this, and with the hindsight now from my previous experience, it will take some time.

- I opened an issue about the syntax and my plans on the 26th of November, so roughly 2 months ago, and nobody commented on the syntax change (#9315)
- We shipped the reading part of the new syntax 1 week ago without bigger discussion
- This PR was ready and approved, and the discussion just started because I disabled the "line to long" check

So it's a little bit sad to me that I wasted so much time implementing this without much more comments on the syntax, and then we decided to change it just a few hours before I wanted to pull the trigger :(

carlocab commented:
>So it's a little bit sad to me that I wasted so much time implementing this

I'm aware, and I feel a bit guilty even sharing my thoughts about it. However, thinking about it more, I think the improvements from a major overhaul might be too marginal to be worth the investment.

Between your option 1 and 2, though, I prefer option 1, as it mirrors more closely how `sha256` lines look elsewhere (and I think that makes things more readable).

MikeMcQuaid commented:
> * and the discussion just started because I disabled the "line to long" check

I'm also sad about this and I apologise for not catching this until now. I think the problem is that I only just noticed that that "line too long" check was going to kick in for multiple lines on every single formula with a bottle and that we're going to be putting important information on the end. I debated whether or not to say anything and I figured: better a bit more pain now (which I can try to be more helpful with) than regretting our decision. I'll happily be overruled on that, though, as I think reordering could achieve some of the same result.

iMichka commented:
I think reordering like this:

```
sha256 cellar: :any_skip_relocation, "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur
```

doest not need a big change, I just need to change the order of writing, and the reading should be fine. As it's a hash and we made sure to not depend on the order (you made me remove all the .first calls, remember? :D), it should work.

I just find it a little bit ugly as the cellar information might be missing on some lines, and the hexdigests will not be aligned anymore.

sjackman commented:
> I think the keys and values for the sha256 hash in the bottle block are the wrong way around...

I've always thought this as well.

I would suggest

``ruby
sha256 cellar: :any_skip_relocation, big_sur: "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f"
```

and please accept my apologies for the late stage bike shedding, Michka, but if we've opened this Pandora's box, that's my two cents.

dawidd6 commented:
> I just find it a little bit ugly as the cellar information might be missing on some lines, and the hexdigests will not be aligned anymore.

Agreed.

SeekingMeaning commented:
This feels a bit wrong to suggest—especially after already adding read support for the new syntax—but what about changing `cellar` to accept a hash?

For example:

```rb
bottle do
  cellar big_sur:  :any_skip_relocation,
         catalina: :any_skip_relocation
  sha256 "1a9fd37585d31ef6f1c61b431bfab4fde9755db5ee081d8eba1c9a358278481f" => :big_sur
  sha256 "99aea0397bae82f956eab6868704b493b99f9160e4c5f8f558937db3e349b96b" => :catalina
end
```

or:

```rb
bottle do
  cellar big_sur: :any_skip_relocation
  cellar catalina: :any_skip_relocation
  sha256 "1a9fd37585d31ef6f1c61b431bfab4fde9755db5ee081d8eba1c9a358278481f" => :big_sur
  sha256 "99aea0397bae82f956eab6868704b493b99f9160e4c5f8f558937db3e349b96b" => :catalina
end
```

iMichka commented:
> Should we move this discussion to #9315?

Let's finish it here.

@SeekingMeaning  regarding your proposal:
```
bottle do
  cellar big_sur: :any_skip_relocation
  cellar catalina: :any_skip_relocation
  sha256 "1a9fd37585d31ef6f1c61b431bfab4fde9755db5ee081d8eba1c9a358278481f" => :big_sur
  sha256 "99aea0397bae82f956eab6868704b493b99f9160e4c5f8f558937db3e349b96b" => :catalina
end
```

this means reimplementing the whole collector logic for these lines too. There is a bunch of logic around the order of the tags in bottle blocks: I think this change would be bigger from a code point of view (at least from what I remember dealing with last time).

iMichka commented:
The same holds for the proposal for @MikeMcQuaid and @carlocab :
```
big_sur cellar: :any_skip_relocation, sha256: "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f"
```

This means creating a "big_sur" function on the fly, and a bunch of functions for each platform: the whole collector logic in the code will be impacted, and this means creating tag specific functions for ALL possible platforms. From my point of view, this change in the code will be massive.

carlocab commented:
I'm happy with option 1:
```ruby
sha256 "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f" => :big_sur, :cellar => :any_skip_relocation
```
This preserves the uniformity of appearance with the `sha256` lines elsewhere and is already readable for anyone who currently looks inside the `bottle` blocks. One will have to look to the end of the line for the useful information currently anyway, and this is an improvement on that by including more useful information where you're already looking.

It would've been nice to put all the interesting information first, but I don't mind chalking this up to something that one would do were they starting from scratch.

MikeMcQuaid commented:
> This feels a bit wrong to suggest—especially after already adding read support for the new syntax, and this PR nearly being merged—but what about changing `#cellar` to accept a hash?

I think it's good to keep together the information for each platform on a line for each platform.

> This means creating a "big_sur" function on the fly, and a bunch of functions for each platform: the whole collector logic in the code will be impacted, and this means creating tag specific functions for ALL possible platforms. From my point of view, this change in the code will be massive.

It's actually relatively easy to do this in Ruby (and I'm happy to be the one to do so, to be clear). Let's focus on what we would ideally have it look like and ignore the implementation for now.

> One has to look to the end of the line for the useful information at the moment anyway, and this is an improvement on that by including more useful information where you're already looking.

The big difference is: this information will be hidden in e.g. the GitHub diff UI without scrolling now.

> It would've been nice to put all the interesting information first, but I don't mind chalking this up as something that one would do were they starting (everything, i.e. `brew` with binary bottles from the ground up) from scratch.

I mean: we are kinda redoing the bottle lines from scratch?

> I would suggest
> 
> ```ruby
> sha256 cellar: :any_skip_relocation, big_sur: "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f"
> ```

This makes sense to me, too. I'd be fine with the line being longer if this is the case because the training information is the least important. 

If I do have to factor in "what's the minimal changes required to make this happen": this is probably my preference. @iMichka what do you think? If you're happy enough with that, I should be able to open a PR today to make this syntax work.

iMichka commented:
I'm fine with that (e.g.)
```
sha256 cellar: :any_skip_relocation, big_sur: "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f"
```

I wont have time to open a PR for this right now, maybe way later today, so if you can beat me to it, I'll let you do it. That PR should only change the reading part to be able to read the new syntax. Once this is done and merged, I'll rebase this PR accordingly.

carlocab commented:
```ruby
sha256 cellar: :any_skip_relocation, big_sur: "a0af7dcbb5c83f6f3f7ecd507c2d352c1a018f894d51ad241ce8492fa598010f"
```
This would make the bottle block much more difficult to read in the (admittedly rather unlikely) event that different versions have different `cellar` values.

iMichka commented:
Both solutions have drawbacks:
- Putting the cellar information on the left makes the hexdigest non-aligned and less readable when different cellar information is used based on the platform (my concerns above)
- Putting it on the right forces you to scroll to see that information if you have a small window, or in the github UI diff.

So, you need to chose which one is less worse.

carlocab commented:
>Putting it on the right forces you to scroll to see that information if you have a small window, or in the github UI diff.

I use a screen on the smaller side (13"), and I usually have enough window space to see the whole `sha256` line in the cellar block with lots of room left over. I'm also partial to unified diffs. And I think the split diffs wrap. If so, then it may not be a matter of scrolling as an issue of reading wrapped lines.

Anyway, you're right about both having drawbacks. I don't mind trusting your judgment on this.

MikeMcQuaid commented:
> I wont have time to open a PR for this right now, maybe way later today, so if you can beat me to it, I'll let you do it. That PR should only change the reading part to be able to read the new syntax. Once this is done and merged, I'll rebase this PR accordingly.

Working on it now 👍🏻

MikeMcQuaid commented:
> Once this is done and merged, I'll rebase this PR accordingly.

Have done this for you too.

iMichka commented:
Thanks. I think you messed something up with the branches here :D

MikeMcQuaid commented:
Oops can you try to reopen?

MikeMcQuaid commented:
Try to change the base branch back to master.

iMichka commented:
> Try to change the base branch back to master.

How do I do that. In git?

I can repush my local branch without your last commit, is that ok for you?

sjackman commented:
Superseded by PR https://github.com/Homebrew/brew/pull/10449
